File: src/Analyzer.hs
module Analyzer (analyze) where

import Type

executeApplication :: Exp -> [Exp] -> Env -> (Exp, Env)

executeApplication (Lambda parameters body) args env =
  bproc $ extendEnv parameters args env
  where
    bproc = analyze body

executeApplication (Primitive f) args env = (f args, env)

analyze :: Exp -> Env -> (Exp, Env)
analyze (Application (operator:operands)) =
  \env ->
     let (app, _) = (fproc env)
         args = map fst (map (\aproc -> aproc env) aprocs)
     in executeApplication app args env
  where
    fproc = analyze operator
    aprocs = map analyze operands

analyze (Sym name) = \env -> (lookupValue name env, env)

analyze (Def (Sym name) exp) = \env -> (Unit, insertValue name exp env)

analyze (If predicate consequent alternative) =
  \env ->
    let ((Boolean p), _) = (pproc env)
    in if p then cproc env else aproc env
  where
    pproc = analyze predicate
    cproc = analyze consequent
    aproc = analyze alternative
  
analyze x = \env -> (x, env)

File: kand/src/kand/analyzer.clj
(ns kand.analyzer
  (:refer-clojure :exclude [name])
  (:require [kand.type :refer :all])
  (:import [kand.type Application If Def Lambda Primitive Symbol Num Unit]))

(defmulti analyze type)

(defmulti execute (fn [v _ _] (type v)))

(defmethod execute Lambda [{:keys [params body]} args env]
  (let [bproc (analyze body)]
    (bproc (merge env (zipmap (map :name params) args)))))

(defmethod execute Primitive [{:keys [f]} args env]
  [(f args) env])

(defmethod execute :default [_ _ env]
  [(->Unit) env])

(defmethod analyze Application [{:keys [exps]}]
  (let [[operator & operands] exps
        fproc (analyze operator)
        aprocs (map analyze operands)]
    #(let [[app] (fproc %)
           args (map first (map (fn [aproc] (aproc %)) aprocs))]
       (execute app args %))))

(defmethod analyze Symbol [{:keys [name]}]
  #(vector (get % name) %))

(defmethod analyze Def [{:keys [name body]}]
  #(vector (->Unit) (assoc % (:name name) body)))

(defmethod analyze If [{:keys [pred t f]}]
  (let [pproc (analyze pred)
        tproc (analyze t)
        fproc (analyze f)]
    #(let [[p] (pproc %)]
       (if p
         (tproc %)
         (fproc %)))))

(defmethod analyze :default [exp]
  #(vector exp %))


